src/go/wsl-helper/pkg/dockerproxy/hyperv.go

```diff
@@ -0,0 +1,130 @@
+//go:build windows
+// +build windows
+
+/*
+Copyright © 2021 SUSE LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dockerproxy
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/linuxkit/virtsock/pkg/hvsock"
+	"golang.org/x/sys/windows/registry"
+)
+
+// Racer is a helper structure to return either a successful result (the GUID of
+// the Hyper-V virtual machine for WSL2), or an error if all attempts have
+// failed.
+type racer struct {
+	result    hvsock.GUID
+	lastError error
+	count     int
+	lock      sync.Locker
+	cond      *sync.Cond
+}
+
+// newRacer constructs a new racer that will return an error after the given
+// number of calls to reject().
+func newRacer(n int) *racer {
+	mutex := &sync.Mutex{}
+	return &racer{
+		result: hvsock.GUIDZero,
+		count:  n,
+		lock:   mutex,
+		cond:   sync.NewCond(mutex),
+	}
+}
+
+// Wait for a result; either the parameter of any resolve() call, or the last
+// error from a reject() call.
+func (r *racer) wait() (hvsock.GUID, error) {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+	for r.count > 0 {
+		r.cond.Wait()
+	}
+	if r.result != hvsock.GUIDZero {
+		return r.result, nil
+	}
+	return hvsock.GUIDZero, r.lastError
+}
+
+// Resolve the racer with the given successful result.
+func (r *racer) resolve(guid hvsock.GUID) {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+	if r.result == hvsock.GUIDZero {
+		r.result = guid
+	}
+	r.count = 0
+	r.cond.Signal()
+}
+
+// Reject the racer with the given unsuccessful result.
+func (r *racer) reject(err error) {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+	r.lastError = err
+	r.count -= 1
+	r.cond.Signal()
+}
+
+// Probe the system to detect the correct VM GUID for the WSL virtual machine.
+// This requires that WSL2 is already running.
+func probeVMGUID(port uint32) (hvsock.GUID, error) {
+	key, err := registry.OpenKey(
+		registry.LOCAL_MACHINE,
+		`SOFTWARE\Microsoft\Windows NT\CurrentVersion\HostComputeService\VolatileStore\ComputeSystem`,
+		registry.ENUMERATE_SUB_KEYS)
```

Comment:
The documented way to do this is via [`HcsEnumerateComputeSystems()`](https://docs.microsoft.com/en-us/virtualization/api/hcs/reference/hcsenumeratecomputesystems), but that requires additional privileges that a normal standard user wouldn't have.

Reply:
Not in this PR, but typo at https://github.com/rancher-sandbox/rancher-desktop/blob/0c3245708401afbe1f45aedf8b5c33612cbb3a18/src/go/wsl-helper/cmd/root.go#L27 "intergation"